lib/howl/command.moon

```diff
@@ -182,48 +83,147 @@ command_completer = ->
   matcher = Matcher items
   (text) -> matcher(text), completion_options
 
-direct_dispatch = (cmd_string) ->
-  return false if not cmd_string or cmd_string.is_blank
-  cmd, cmd_name, text = parse_cmd cmd_string
-  cmd or= resolve_command cmd_string
-  return false unless cmd
-  input = load_input cmd.input, text
-  return false if input
-  cmd!
-  true
-
-run = (cmd_string = nil) ->
-  return if direct_dispatch cmd_string
-
-  cmd_completer = nil
-  state = State!
-
-  cmd_input =
-    title: 'Command'
-    should_complete: (_, readline) -> state\should_complete readline
-    close_on_cancel: (_, readline) -> state\close_on_cancel readline
-    update: (_, text, readline) -> state\update text, readline
-    on_selection_changed: (_, item, readline) -> state\on_selection_changed item, readline
-    on_completed: (_, item, readline) -> state\on_completed item, readline
-    on_cancelled: (_, readline) -> state\on_cancelled readline
-    on_submit: (_, text, readline) -> state\on_submit text, readline
-    go_back: (_, readline) -> state\go_back readline
-    keymap: state.keymap
-
-    complete: (_, text, readline) ->
-      if state.cmd
-        return state\complete(text, readline)
+get_command_items = ->
+  cmd_names = names!
+  bindings = command_bindings!
+
+  table.sort cmd_names
+  items = {}
+  for name in *cmd_names
+    def = commands[name]
+    desc = def.description
+    if def.alias_for
+      desc = "(Alias for #{def.alias_for})"
+      desc = "[deprecated] #{desc}" if def.deprecated
+    binding = bindings[name] or ''
+    append items, { name, binding, desc }
+
+  return items
+
+class CommandRunner
+  run: (@finish, opts={}) =>
+    @command_line = howl.app.window.command_line
+    cmd = nil
+    cmdname = ''
+    @command_line.prompt = ':'
+    @command_line.title = 'Command'
+    if opts.directory
+      @command_line.directory = opts.directory
+    if opts.cmd_string
+      @run_command opts.cmd_string
+
+  run_command: (cmd_string) =>
+    cmd, cmdname, text = parse_cmd cmd_string
+
+    if cmd
+      @command_line\write_spillover text
+    else
+      cmd = resolve_command cmd_string
+      cmdname = cmd and cmd.name
+
+    if not cmd
+      @command_line.text = cmd_string
+      log.error 'No such command'
+      return
+
+    @command_line\clear!
+    @command_line.prompt = markup.howl "<prompt>:</><commandname>#{cmdname}</> "
+
+    local results
+
+    ok, error = pcall ->
+      results = table.pack howl.app.window.command_line\run cmd
+
+    if not ok
+      log.error error
+
+    self.finish results and unpack results
+
+  on_update: (text) =>
+    if text\find ' '
```

Comment:
I interested in this, since it seems like it would mean that we'll never be able to specify arguments with spaces in them? Anyway, I experimented a little bit and encountered an issue. In my case I opened the command prompt, typed 'ctrl' and entered space. This caused the current command in the completion list to be run (`view-close` in my case), which does not seem desirable.